### PR TITLE
TS-4705: Proposal: NetVC Context

### PR DESCRIPTION
**Pull request by oknet**
Goal 1st:
In the NetVConnection, we have get_local_addr() and get_remote_addr() methods.
Also have members local_addr, remote_addr and netvc->con.addr.

Thus, we should using netvc->con.addr or remote_addr to replace member server_addr in UnixNetVConnection.

Goal 2nd:
SSLNetVConnection has member sslClientConnection with 2 methods setSSLClientConnection() and getSSLClientConnection() to indictor ATS is a client or server in a SSL session.

To abstract above two goals, I'm design the netvc context function.

As a proxy, there has two side: client side ( Client <-> Proxy ) and server side ( Proxy <-> Server ). With the netvc context funtion to indicate which side the NetVC working on.

Goal 3rd:
Fix a minor bug in NetAccept::do_blocking_accept, call to check_emergency_throttle(con) first then allocate vc.

Goal 4th:
NetAccept Optimize, remove dup code, etc...

### CODE REVIEW

File: iocore/net/UnixNetAccept.cc [hunk 1]
```diff
@@ -117,11 +117,10 @@ net_accept(NetAccept *na, void *ep, bool blockable)
     vc->id = net_next_connection_number();
     vc->con.move(con);
     vc->submit_time = Thread::get_hrtime();
-    ats_ip_copy(&vc->server_addr, &vc->con.addr);
     vc->mutex   = new_ProxyMutex();
     vc->action_ = *na->action_;
     vc->set_is_transparent(na->server.f_inbound_transparent);
-    vc->closed = 0;
```
Comment (oknet) [inline iocore/net/UnixNetAccept.cc]: closed is set to 0 in constructor function. remove dup code.

File: iocore/net/UnixNetAccept.cc [hunk 2]
```diff
@@ -281,29 +280,28 @@ NetAccept::do_blocking_accept(EThread *t)
...
 
+    // con.close() called if return true.
+    if (check_emergency_throttle(con)) {
+      return -1;
+    }
+
     // Use 'NULL' to Bypass thread allocator
     vc = (UnixNetVConnection *)this->getNetProcessor()->allocate_vc(NULL);
     if (unlikely(!vc || shutdown_event_system == true)) {
-      con.close();
```
Comment (oknet) [inline iocore/net/UnixNetAccept.cc]: con.close() is called in destructor function.

File: iocore/net/UnixNetAccept.cc [hunk 3]
```diff
@@ -281,29 +280,28 @@ NetAccept::do_blocking_accept(EThread *t)
...
       return -1;
     }
 
-    vc->con                 = con;
+    NET_SUM_GLOBAL_DYN_STAT(net_connections_currently_open_stat, 1);
     vc->id                  = net_next_connection_number();
-    vc->from_accept_thread  = true;
```
Comment (oknet) [inline iocore/net/UnixNetAccept.cc]: from_accept_thread is set to true in allocate_vc(NULL). remove dup code.

**Comment by oknet** [comment]
This is a rebase version of #642

**Comment by zwoop** [comment]
[approve ci].

**Comment by atsci** [comment]
Linux build *failed*! See https://ci.trafficserver.apache.org/job/Github-Linux/263/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/368/ for details.

File: iocore/net/UnixNetAccept.cc [hunk 4]
```diff
@@ -443,19 +430,28 @@ NetAccept::acceptFastEvent(int event, void *ep)
       goto Lerror;
     }
 
+    vc = (UnixNetVConnection *)this->getNetProcessor()->allocate_vc(e->ethread);
+    if (!vc) {
+      goto Ldone;
```
Comment (shinrich) [inline iocore/net/UnixNetAccept.cc]: Might as well do con.close() here.  Might not be needed now, but might close a leak in the future.  This is the error case anyway, shouldn't micro-optimize here.

**Comment by shinrich** [comment]
Overall I like this idea.  Unifying the multiple copies of addresses is great.  They have only been a source of inconsistencies and pain for me.  Adding the general Context type also seems good.

My only concern is that the diffs in UnixNetAccept are kind of noisy.  I'll take another more care pass later to make sure everything lines up.

File: iocore/net/UnixNetAccept.cc (hunk 4 above)
Comment (oknet) [inline iocore/net/UnixNetAccept.cc]: con.close() is called in destructor function.

**Comment by oknet** [comment]
@shinrich the "noisy" in UnixNetAccept, let me explain, there are 3 area to create a new netvc and init its members but disorder, I'm try to make it ordered as below in these 3 area.

    NET_SUM_GLOBAL_DYN_STAT(net_connections_currently_open_stat, 1); 
    vc->id = net_next_connection_number();
    vc->con.move(con);
    vc->submit_time = Thread::get_hrtime();
    vc->mutex  = new_ProxyMutex();
    // no need to set vc->action_
    vc->set_is_transparent(server.f_inbound_transparent);
    vc->options.packet_mark = packet_mark;
    vc->options.packet_tos  = packet_tos;
    vc->apply_options();
    vc->set_context(Net_VConnection_C2P);
    SET_CONTINUATION_HANDLER(vc, (NetVConnHandler)&UnixNetVConnection::mainEvent);

    // set thread and nh as acceptEvent does
    vc->thread = e->ethread;
    vc->nh = get_NetHandler(e->ethread);

**Comment by zwoop** [comment]
@shinrich  Should we continue with this PR? @oknet there are now merge conflicts, can you please rebase (and remember to re-run clang-format too :).

**Comment by oknet** [comment]
@zwoop rebase & clang-format

**Comment by zwoop** [comment]
[approve ci]

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/482/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/379/ for details.

File: iocore/net/I_NetVConnection.h [hunk 5]
```diff
@@ -40,6 +40,13 @@
 #define SSL_EVENT_SERVER 0
 #define SSL_EVENT_CLIENT 1
 
+// Indicator the context for a NetVConnection
+typedef enum {
+  Net_VConnection_UNSET,
+  Net_VConnection_C2P, // Client <--> ATS, Client-Side
+  Net_VConnection_P2S, // ATS <--> Server, Server-Side
```
Comment (bryancall) [inline iocore/net/I_NetVConnection.h]: Instead of C2P and P2S, NET_CONNECTION_IN,  NET_CONNECTION_OUT, and  NET_CONNECTION_UNSET would be better.  We normally uppercase enums.

File: iocore/net/UnixNetVConnection.cc [hunk 6]
```diff
@@ -1296,7 +1294,7 @@ UnixNetVConnection::connectUp(EThread *t, int fd)
   }
 
   if (fd == NO_FD) {
-    res = con.connect(&server_addr.sa, options);
+    res = con.connect(NULL, options);
```
Comment (masaori335) [inline iocore/net/UnixNetVConnection.cc]: Should be `&con.addr.sa` ? `&server_addr.sa` was always `NULL` in here ?

File: iocore/net/UnixConnection.cc [hunk 7]
```diff
@@ -320,14 +320,15 @@ Connection::connect(sockaddr const *target, NetVCOptions const &opt)
...
+  if (target != NULL)
+    this->setRemote(target);
 
   // apply dynamic options with this.addr initialized
   apply_options(opt);
 
   cleaner<Connection> cleanup(this, &Connection::_cleanup); // mark for close until we succeed.
 
-  res = ::connect(fd, target, ats_ip_size(target));
+  res = ::connect(fd, &this->addr.sa, ats_ip_size(&this->addr.sa));
```
Comment (oknet) [inline iocore/net/UnixConnection.cc]: @masaori335 con.connect() always using addr.sa as ::connect() target addr.
set addr.sa with target if target is not NULL.

File: iocore/net/UnixNetVConnection.cc (hunk 6 above)
Comment (oknet) [inline iocore/net/UnixNetVConnection.cc]: In UnixNetProcessor::connect_re_internal(), the con.addr.sa set with target addr.
The member server_addr is removed from NetVConnection in this PR.

File: iocore/net/UnixConnection.cc (hunk 7 above)
Comment (masaori335) [inline iocore/net/UnixConnection.cc]: @oknet Ah, got it. Thanks!

**Comment by masaori335** [comment]
BTW, which is JIRA ticket for this? And would you use the ticket number in commit message?
Our commit log format requires it.
https://cwiki.apache.org/confluence/display/TS/CommitPolicies

**Comment by oknet** [comment]
@masaori335 JIRA issue created. 
BTW what is the proposal review progress ?
Is it same as the bug report progress ? create JIRA issue then create PR ?

**Comment by masaori335** [comment]
@oknet AFAIK, it's same. Please set type of JIRA 'improvement'. Currently all code changes are tracked by JIRA.

**Comment by oknet** [comment]
@bryancall update the enums. 
```
  NET_VCONNECTION_UNSET,
  NET_VCONNECTION_IN, // Client <--> ATS, Client-Side
  NET_VCONNECTION_OUT, // ATS <--> Server, Server-Side
```

**Comment by shinrich** [comment]
Looks good to me!

**Comment by oknet** [comment]
rebase after TS-4779

File: iocore/net/I_NetVConnection.h [hunk 8]
```diff
@@ -566,6 +573,21 @@ class NetVConnection : public VConnection
     is_transparent = state;
   }
 
+  /// Set the context and return context value
+  NetVConnectionContext_t
+  set_context(NetVConnectionContext_t context)
```
Comment (jpeach) [inline iocore/net/I_NetVConnection.h]: The return value of this is never used, so can we remove it? If the expectation is that the context is only set once, can we either
    - do it in the constructor and remove this
    - assert that it is unset here.

File: iocore/net/I_NetVConnection.h [hunk 9]
```diff
@@ -566,6 +573,21 @@ class NetVConnection : public VConnection
...
+  {
+    if (NET_VCONNECTION_UNSET == netvc_context)
+      netvc_context = context;
+    return netvc_context;
+  }
+  /// Get the context.
+  NetVConnectionContext_t
+  get_context() const
+  {
```
Comment (jpeach) [inline iocore/net/I_NetVConnection.h]: Can we assert that the context is never NET_VCONNECTION_UNSET?

Comment (jpeach) [inline iocore/net/I_NetVConnection.h]: I know we don't have strong conventions, but I'd prefer to aim for ``context()`` and ``set_context()`` as the member functions. The former reads better than ``get_context()``. As noted above I'd like to try to eliminate ``set_context()``.

File: iocore/net/P_SSLNetVConnection.h [hunk 10]
```diff
@@ -101,26 +101,13 @@ class SSLNetVConnection : public UnixNetVConnection
...
-  }
-
-  void
   setSSLSessionCacheHit(bool state)
   {
     sslSessionCacheHit = state;
   }
 
-  bool
+  virtual bool
```
Comment (jpeach) [inline iocore/net/P_SSLNetVConnection.h]: Why did you add this?

Comment (oknet) [inline iocore/net/P_SSLNetVConnection.h]: sorry, It is a rebase & merge mistaken.

File: iocore/net/I_NetVConnection.h (hunk 9 above)
Comment (oknet) [inline iocore/net/I_NetVConnection.h]: Actually, I did reference the existing functions to name these two functions, such as "set_remote_addr" and "get_remote_addr".

File: iocore/net/I_NetVConnection.h (hunk 8 above)
Comment (oknet) [inline iocore/net/I_NetVConnection.h]: The return value tells you whether it sets successful or not.
Yes, the context is only set once and will not change.
- Due to the VC is create by Allocator, It can not be set in the constructor.
- Assert here and set return type to void. ( update later )

**Comment by oknet** [comment]
Depend on @jpeach's suggestion:
- the return type of set_context is void now.
- add ink_assert inside set_context()

Also:
- set NET_VCONNECTION_UNSET to 0 in the enum definition.
- add comments for set_context() and get_context().

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/673/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/777/ for details.

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/726/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/830/ for details.

**Comment by oknet** [comment]
bugfix: In do_blocking_accept(), we should check the con.fd == NO_FD while we get return from check_emergency_throttle().

File: iocore/net/UnixNetAccept.cc [hunk 11]
```diff
@@ -281,29 +280,30 @@ NetAccept::do_blocking_accept(EThread *t)
       return -1;
     }
 
+    if (check_emergency_throttle(con)) {
+      // `con' could be closed if there is hyper emergency
+      if (con.fd == NO_FD) {
+        return 0;
+      }
+    }
+
```
Comment (oknet) [inline iocore/net/UnixNetAccept.cc]: @jpeach , Here is a similar bug as PR #1033

**Comment by atsci** [comment]
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/739/ for details.

**Comment by atsci** [comment]
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/844/ for details.
